pyasn1/codec/ber/decoder.py

```diff
@@ -16,35 +19,71 @@
 from pyasn1.type import univ
 from pyasn1.type import useful
 
-__all__ = ['decode']
+
+__all__ = ['decode', 'decodeStream']
 
 LOG = debug.registerLoggee(__name__, flags=debug.DEBUG_DECODER)
 
 noValue = base.noValue
 
 
+def asStream(substrate):
+    # TODO: Apply sparingly or remove
+    # :type substrate: Union[bytes, BytesIO, BufferedReader, OctetString]
+    # :rtype: BufferedReader
+    if isinstance(substrate, univ.OctetString):
+        substrate = substrate.asOctets()
+    if isinstance(substrate, bytes):
+        substrate = BytesIO(substrate)
+    if isinstance(substrate, BytesIO):
+        return BufferedReader(substrate)
+    if isinstance(substrate, BufferedReader):
+        return substrate
+    raise ValueError("%s Cannot be converted to stream" % substrate.__class__)
+
+
+def popSubstream(substrate, length):
+    content = substrate.read(length)
+    return asStream(content)
+
+
+def isEmpty(substrate):
+    # :type substrate: Union[bytes, BytesIO, BufferedReader]
+    # :rtype: bool
+    if isinstance(substrate, bytes):
+        return not bool(substrate)
+    else:
+        return not bool(asStream(substrate).peek(1))
+
+
 class AbstractDecoder(object):
     protoComponent = None
 
     def valueDecoder(self, substrate, asn1Spec,
                      tagSet=None, length=None, state=None,
                      decodeFun=None, substrateFun=None,
                      **options):
+        # Note: return the value only
         raise error.PyAsn1Error('Decoder not implemented for %s' % (tagSet,))
 
     def indefLenValueDecoder(self, substrate, asn1Spec,
                              tagSet=None, length=None, state=None,
                              decodeFun=None, substrateFun=None,
                              **options):
+        # Note: return the value only
         raise error.PyAsn1Error('Indefinite length mode decoder not implemented for %s' % (tagSet,))
 
 
 class AbstractSimpleDecoder(AbstractDecoder):
     @staticmethod
     def substrateCollector(asn1Object, substrate, length):
-        return substrate[:length], substrate[length:]
+        # :type substrate: BytesIO
+        # :type length: int
+        # :rtype: object
+        return substrate.read(length)
```

Comment:
What happens if input of size `length` is not presently available? Will it block? What would be great, in my opinion, is if we yield instead of blocking. That would let the application to do other things while sufficient amount of input becomes available.

Reply:
I guess it would block. However, I am not sure how to "let the application do other things" (with this design). If you know, I would be glad to learn :-)

Reply:
I've commented on that matter in #176